### PR TITLE
smt2_solver: arrays

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2/smt2_parser.cpp [hunk 1]
```diff
@@ -644,6 +644,31 @@ exprt smt2_parsert::function_application()
       {
         return binary(ID_implies, op);
       }
+      else if(id == "select")
+      {
+        // array index
+        if(op.size() != 2)
+          throw error("index takes two operands");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_parser.cpp]: Should this say "select"?

File: src/solvers/smt2/smt2_parser.cpp [hunk 2]
```diff
@@ -644,6 +644,31 @@ exprt smt2_parsert::function_application()
...
+
+        if(op[0].type().id() != ID_array)
+          throw error("index expects array operand");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_parser.cpp]: How about "select expects array as first operand"

File: src/solvers/smt2/smt2_parser.cpp [hunk 3]
```diff
@@ -644,6 +644,31 @@ exprt smt2_parsert::function_application()
...
+      {
+        // array update
+        if(op.size() != 3)
+          throw error("store takes three operands");
+
+        if(op[0].type().id() != ID_array)
+          throw error("store expects array operand");
+
+        if(to_array_type(op[0].type()).subtype() != op[2].type())
+          throw error("store expects value that matches array elements");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_parser.cpp]: "... matches array element type"?

**Review by tautschnig — APPROVED** [review]

File: src/solvers/smt2/smt2_parser.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: sure

File: src/solvers/smt2/smt2_parser.cpp (hunk 3 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: of course

File: src/solvers/smt2/smt2_parser.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: done

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4f485019e8aff253db0ffc67592c97073b960664).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93368956
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: e62c0ea906c1f9b9d648d6ea9f2fa6e7441b6529).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93398589
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**